decentraland/unity-renderer
# refactor: hud controller

Fixes #589 

## What does this PR change?

This PR implements the first stage of this ADR: https://github.com/decentraland/adr/blob/main/docs/ADR-29-refactor-hud-controller.md

- We refactor `HUDController` applying an abstract factory pattern to allow us to override the behaviour of the `HUDs` in `explorer-desktop`.
- Now `HUDController` is no longer a `MonoBehavior`, the life cycle is in new Environment Context named `HUDContext`.
- We refactor a lot of `.asmdef` due to a cycled dependency on the `HUDs`. So now every `Environment ContextFactory` has its own `.asmdef`, and with that, a many `.asmdef` had to change.
- Now `Main` is in a new folder, with its own `.asmdef`.
- Many functions were changed to `virtual functions` to be overridable in `explorer-desktop`, and the same with the visibility from `private` to `protected`.
- We implemented in `explorer-desktop` an `Exit button` to validate if this refactor is what we need.

## How to test the changes?

1. Go to: https://play.decentraland.zone/index.html?renderer=urn:decentraland:off-chain:renderer-artifacts:refactor/hud-controller
2. Check if the HUD's are OK.
3. Go to: decentraland/explorer-desktop#3 and test the `Exit button` (it use this refactor)

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
